src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/LspFileBasedProgramWorkspace.cs

```diff
@@ -0,0 +1,323 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Basic.Reference.Assemblies;
+using Microsoft.CodeAnalysis.Collections;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Features.Workspaces;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageServer.Handler;
+using Microsoft.CodeAnalysis.LanguageServer.Handler.DebugConfiguration;
+using Microsoft.CodeAnalysis.LanguageServer.HostWorkspace;
+using Microsoft.CodeAnalysis.MetadataAsSource;
+using Microsoft.CodeAnalysis.MSBuild;
+using Microsoft.CodeAnalysis.ProjectSystem;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Threading;
+using Microsoft.CodeAnalysis.Workspaces.ProjectSystem;
+using Microsoft.CommonLanguageServerProtocol.Framework;
+using Microsoft.Extensions.Logging;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer;
+
+/// <summary>
+/// Workspace which handles file-based programs ('dotnet run app.cs' scenario).
+/// </summary>
+internal sealed class LspFileBasedProgramWorkspace : Workspace, ILspService, ILspWorkspace, ILspFileBasedProgramWorkspace
+{
+    /// <summary>
+    /// A single gate for code that is adding work to <see cref="_projectsToLoadAndReload" />. This is just we don't have code simultaneously trying to load and unload solutions at once.
+    /// </summary>
+    private readonly SemaphoreSlim _gate = new SemaphoreSlim(initialCount: 1);
+
+    // TODO: should we use the LanguageServerProjectSystem here instead?
+    private readonly AsyncBatchingWorkQueue<(Uri uri, SourceText documentText, string languageId)> _projectsToLoadAndReload; // TODO: do we need to introduce a shutdown cancellation token for this?
+
+    private readonly ConcurrentDictionary<string, List<LoadedProject>> _loadedProjects = [];
```

Comment:
we have a _gate, but also a ConcurrentDict?  needs docs explaining the threading mmodel.

Reply:
planning to delete this and reuse the LanguageServerProjectSystem.